File: LC54.java
// 54. Spiral Matrix


class Solution {
    int[][] dir = {{0,1}, {1,0}, {0, -1}, {-1,0}}; //RBLT
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> out = new ArrayList<>();    
        traverseSpiral(matrix , 0, 0, out ,0);  
        return out;
    }

    public void traverseSpiral(int[][] m, int r, int c,  List<Integer> out ,int dir){
        if(r < 0 || r >= m.length || c < 0 || c >= m[0].length || m[r][c] < -101 ) return;
        out.add(m[r][c]);
        m[r][c] = -200;
        int[] d = this.dir[dir];
         if(r + d[0] < 0 || r + d[0] >= m.length || c + d[1] < 0 || c + d[1] >= m[0].length || m[r + d[0]][c + d[1]] < -101 ){
             //change direction
             dir = (dir+1)%4;
             d = this.dir[dir];
         }
        traverseSpiral(m , r + d[0], c + d[1], out, dir);
    }
}

File: LC54.go
// 54. Spiral Matrix

func spiralOrder(matrix [][]int) []int {
	dir := [][]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
	out := make([]int, len(matrix)*len(matrix[0]))
	i := 0
	d := 0
	r, c := 0, 0
	for i < len(matrix)*len(matrix[0]) {
		out[i] = matrix[r][c]
		i++
		matrix[r][c] = -1000
		r += dir[d][0]
		c += dir[d][1]
		if i >= len(matrix)*len(matrix[0]) {
			break
		}
		for c < 0 || r < 0 || c >= len(matrix[0]) || r >= len(matrix) || matrix[r][c] == -1000 {
			r -= dir[d][0]
			c -= dir[d][1]
			d = (d + 1) % 4
			r += dir[d][0]
			c += dir[d][1]
		}
	}

	return out
}